### PR TITLE
Nvfuser code bump 11 5

### PR DESCRIPTION
**Pull request by jjsjann123**
nvfuser code update:
1. Tuning heuristics on schedulers for reduction/normalization kernels;
2. bfloat16 on IO tensor support;
3. Refactored memory format support, now we can support dimension collapsing with non-coherent input tensors with different memory format. e.g. channels last tensor input to batch normalization. Note that we are currently limiting memory format to only Contiguous and Channels last;
4. Refactored nvfuser graph partitioning in `graph_fuser.cpp`, separated node merge and profile node API. Updated `profiling_record.cpp`.

Things that are reverted from our local branch:
1. changes on some entries in autodiff
2. aten::gelu with approximation
3. native_dropout(_backward)

### CODE REVIEW

**Comment by jjsjann123** [comment]
errr, looks like there's some cpp test failure: https://github.com/pytorch/pytorch/runs/4123460077?check_suite_focus=true

probably arch specific. Let me see if I can grab a Maxwell anywhere...

**Comment by jjsjann123** [comment]
Disabled these failing tests in MAXWELL for the time being https://github.com/pytorch/pytorch/runs/4123460077?check_suite_focus=true. I'll double check that with the team on Monday.

Looks like MSVC doesn't like this function: https://github.com/pytorch/pytorch/pull/67943/files#diff-5e44e2a0e9a686349e4152cd3f1255ae55c3c208b22112f194144d8c6612f61aR953-R957

```
2021-11-06T01:30:40.0128999Z C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29333\include\xutility(4393): error C2280: 'std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>> &std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>>::operator =(const std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>> &)': attempting to reference a deleted function
2021-11-06T01:30:40.0132718Z C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29333\include\memory(2687): note: see declaration of 'std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>>::operator ='
2021-11-06T01:30:40.0136597Z C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29333\include\memory(2687): note: 'std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>> &std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>>::operator =(const std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>> &)': function was explicitly deleted
2021-11-06T01:30:40.0140782Z C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29333\include\vector(1127): note: see reference to function template instantiation '_OutIt *std::_Copy_unchecked<_Iter,std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>>*>(_InIt,_InIt,_OutIt)' being compiled
2021-11-06T01:30:40.0142336Z         with
2021-11-06T01:30:40.0142562Z         [
2021-11-06T01:30:40.0143514Z             _OutIt=std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>> *,
2021-11-06T01:30:40.0145703Z             _Iter=std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>> *,
2021-11-06T01:30:40.0147361Z             _InIt=std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>> *
2021-11-06T01:30:40.0148268Z         ]
2021-11-06T01:30:40.0149919Z C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29333\include\vector(1142): note: see reference to function template instantiation 'void std::vector<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr,std::allocator<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr>>::_Assign_range<_Iter>(_Iter,_Iter,std::forward_iterator_tag)' being compiled
2021-11-06T01:30:40.0151522Z         with
2021-11-06T01:30:40.0151743Z         [
2021-11-06T01:30:40.0152685Z             _Iter=std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>> *
2021-11-06T01:30:40.0153615Z         ]
2021-11-06T01:30:40.0155234Z C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29333\include\vector(1142): note: see reference to function template instantiation 'void std::vector<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr,std::allocator<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr>>::_Assign_range<_Iter>(_Iter,_Iter,std::forward_iterator_tag)' being compiled
2021-11-06T01:30:40.0156772Z         with
2021-11-06T01:30:40.0157004Z         [
2021-11-06T01:30:40.0158334Z             _Iter=std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>> *
2021-11-06T01:30:40.0159854Z         ]
2021-11-06T01:30:40.0163351Z C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29333\include\vector(1153): note: see reference to function template instantiation 'void std::vector<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr,std::allocator<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr>>::assign<std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>>*,0>(_Iter,_Iter)' being compiled
2021-11-06T01:30:40.0165643Z         with
2021-11-06T01:30:40.0165881Z         [
2021-11-06T01:30:40.0167207Z             _Iter=std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>> *
2021-11-06T01:30:40.0168785Z         ]
2021-11-06T01:30:40.0172315Z C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29333\include\vector(1153): note: see reference to function template instantiation 'void std::vector<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr,std::allocator<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr>>::assign<std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>>*,0>(_Iter,_Iter)' being compiled
2021-11-06T01:30:40.0176186Z         with
2021-11-06T01:30:40.0176561Z         [
2021-11-06T01:30:40.0178223Z             _Iter=std::unique_ptr<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry,std::default_delete<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::Entry>> *
2021-11-06T01:30:40.0179987Z         ]
2021-11-06T01:30:40.0183433Z C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29333\include\vector(1150): note: while compiling class template member function 'void std::vector<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr,std::allocator<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr>>::_Copy_assign(const std::vector<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr,std::allocator<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr>> &,std::false_type)'
2021-11-06T01:30:40.0188117Z C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29333\include\vector(1169): note: see reference to function template instantiation 'void std::vector<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr,std::allocator<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr>>::_Copy_assign(const std::vector<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr,std::allocator<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr>> &,std::false_type)' being compiled
2021-11-06T01:30:40.0192431Z C:\actions-runner\_work\pytorch\pytorch\torch/csrc/jit/codegen/cuda/executor_utils.h(236): note: see reference to class template instantiation 'std::vector<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr,std::allocator<torch::jit::fuser::cuda::executor_utils::caching::ExecutorCompileTimeInfoCache::EntryOwningPtr>>' being compiled
2021-11-06T01:30:40.0194270Z Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29337 for x64
2021-11-06T01:30:40.0195117Z Copyright (C) Microsoft Corporation.  All rights reserved.
```

Things looks to be moved properly, but MSVC still complains, and strangely I can't repro this with godbolt neither https://godbolt.org/z/rbT9PKnPd

**Comment by jjsjann123** [comment]
So windows build was with exported class containing `std::vector<std::unique_ptr<>>`: https://godbolt.org/z/3E4e8T1P1

got rid of the export since I don't think we need to expose that. :crossed_fingers:

**Comment by jjsjann123** [comment]
rocm failure doesn't look to be related: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.3.1-py3.6-test2/9104/console

clang-tidy is complaining about missing file, which was removed/renamed in this PR: https://github.com/pytorch/pytorch/runs/4145667458?check_suite_focus=true

I'm gonna try upstream merge to see if they go away.

File: torch/csrc/jit/api/function_impl.h [hunk 1]
```diff
@@ -133,7 +133,9 @@ struct TORCH_API GraphFunction : public Function {
   mutable std::recursive_mutex compile_mutex;
 
   // executor_[0] - autocast off
-  // executor_[1] - autocast on
+  // executor_[1] - autocast cpu on
```
Comment (jjsjann123) [inline torch/csrc/jit/api/function_impl.h:136]: fixing some comment.

**Comment by jjsjann123** [comment]
This PR is ready to be imported to test internal build.

**Comment by dzhulgakov** [comment]
rocm test seems flaky, verified with a re-run: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.3.1-py3.6-test2/9849/

**Comment by jjsjann123** [comment]
Failing that I saw in the CI seems to be numerical, re-run in the link shows it's flaky as well. I don't see this PR contributing to that. Is there anything issues that we should be looking at for this PR?

```
00:57:18 ======================================================================
00:57:18 FAIL [0.441s]: test_nadam (__main__.TestOptim)
00:57:18 ----------------------------------------------------------------------
00:57:18 Traceback (most recent call last):
00:57:18   File "test_optim.py", line 630, in test_nadam
00:57:18     lambda weight, bias: optimizer(
00:57:18   File "test_optim.py", line 231, in _test_basic_cases
00:57:18     make_two_arg_constructor(constructor, maximize),
00:57:18   File "test_optim.py", line 209, in _test_state_dict
00:57:18     self.assertEqual(bias, bias_cuda)
00:57:18   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1928, in assertEqual
00:57:18     super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
00:57:18 AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1.3e-06 and atol=1e-05, found 1 element(s) (out of 10) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 1.9073486328125e-05 (-1.0852036476135254 vs. -1.0851845741271973), which occurred at index 1.
00:57:18 
00:57:18 ----------------------------------------------------------------------
```

**Pull request closed without merging**